feat: add FiddleCube guard as a safety provider

kaushik-himself opened the pull request:
# What does this PR do?
Adds FiddleCube guard as a remote safety provider.

[//]: # (If resolving an issue, uncomment and update the line below)
[//]: # (Closes #[issue-number])

## Test Plan
### Setup

Install conda 

Setup a conda environment and activate it:
```bash
conda create -n llama_stack python=3.11 -y 
source $(conda info --base)/etc/profile.d/conda.sh && conda activate llama_stack
```

### Apply Diff
```bash
diff --git a/llama_stack/templates/together/build.yaml b/llama_stack/templates/together/build.yaml
index 90ee5bc..16a9dd3 100644
--- a/llama_stack/templates/together/build.yaml
+++ b/llama_stack/templates/together/build.yaml
@@ -9,7 +9,7 @@ distribution_spec:
     - remote::chromadb
     - remote::pgvector
     safety:
-    - inline::llama-guard
+    - remote::fiddlecube
     agents:
     - inline::meta-reference
     telemetry:
diff --git a/llama_stack/templates/together/run.yaml b/llama_stack/templates/together/run.yaml
index c2afd98..9eddc6b 100644
--- a/llama_stack/templates/together/run.yaml
+++ b/llama_stack/templates/together/run.yaml
@@ -29,8 +29,8 @@ providers:
         namespace: null
         db_path: ${env.SQLITE_STORE_DIR:~/.llama/distributions/together}/faiss_store.db
   safety:
-  - provider_id: llama-guard
-    provider_type: inline::llama-guard
+  - provider_id: fiddlecube
+    provider_type: remote::fiddlecube
     config: {}
   agents:
   - provider_id: meta-reference
@@ -144,7 +144,8 @@ models:
   provider_id: sentence-transformers
   model_type: embedding
 shields:
-- shield_id: meta-llama/Llama-Guard-3-8B
+- shield_id: fiddlecube
+  provider_id: fiddlecube
 vector_dbs: []
 datasets: []
 scoring_fns: []
diff --git a/llama_stack/templates/together/together.py b/llama_stack/templates/together/together.py
index b7ac130..71ddfee 100644
--- a/llama_stack/templates/together/together.py
+++ b/llama_stack/templates/together/together.py
@@ -28,7 +28,7 @@ def get_distribution_template() -> DistributionTemplate:
     providers = {
         "inference": ["remote::together"],
         "vector_io": ["inline::faiss", "remote::chromadb", "remote::pgvector"],
-        "safety": ["inline::llama-guard"],
+        "safety": ["inline::llama-guard", "remote::fiddlecube"],
         "agents": ["inline::meta-reference"],
         "telemetry": ["inline::meta-reference"],
         "eval": ["inline::meta-reference"],
@@ -59,7 +59,9 @@ def get_distribution_template() -> DistributionTemplate:
         config=SentenceTransformersInferenceConfig.sample_run_config(),
     )
 
-    core_model_to_hf_repo = {m.descriptor(): m.huggingface_repo for m in all_registered_models()}
+    core_model_to_hf_repo = {
+        m.descriptor(): m.huggingface_repo for m in all_registered_models()
+    }
     default_models = [
         ModelInput(
             model_id=core_model_to_hf_repo[m.llama_model],

  ```

### Build and Run
```bash
# the following commands need to be run in the conda env created above.
pip install -e .
LLAMA_MODELS_DIR=../llama-models LLAMA_STACK_DIR=$(pwd) llama stack build --image-type conda --template together
llama stack run together 
```

### Test
```bash
curl -X 'POST' \
  'http://localhost:8321/v1/safety/run-shield' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "shield_id": "fiddlecube",
  "messages": [
    {
      "role": "user",
      "content": "Who won the Super Bowl?",
      "context": "Philadelphia Eagles"
    },
    {
      "role": "system",
      "content": "You are a helpful AI assistant"
    },
    {
      "role": "tool",
      "call_id": "t1924",
      "tool_name": "brave_search",
      "content": "Who won the Super Bowl?"
    },
    {
      "role": "assistant",
      "content": "The Eagles won the Super Bowl this year!",
      "stop_reason": "end_of_turn",
      "tool_calls": [
        {
          "call_id": "t1924",
          "tool_name": "brave_search",
          "arguments": {
            "additionalProp1": "Philadelphia",
            "additionalProp2": "Eagles",
            "additionalProp3": "2025"
          }
        }
      ]
    }
  ],
  "params": {}
}'
```

<b>Response</b>
```json
{
  "violation": null
}
```

leseb reviewed (commented):
Thanks for your contribution:

* Could you please update the CHANGELOG.md to reflect this new feature?
* Should we update some distro examples with FiddleCube?
* Any doc update to reflect FiddleCube's availability to users?
* What's a good provider to test with?

Thanks!

kaushik-himself commented:
> Thanks for your contribution:
> 
> * Could you please update the CHANGELOG.md to reflect this new feature?
> 

I don't see any CHANGELOG.md file in this repo. Am I missing something?

leseb commented:
> > Thanks for your contribution:
> > 
> > * Could you please update the CHANGELOG.md to reflect this new feature?
> 
> I don't see any CHANGELOG.md file in this repo. Am I missing something?

My bad, I just noticed that it was removed in https://github.com/meta-llama/llama-stack/commit/e6c9f2a4856192d6cb57a038d98d21a253c4319a. Forget about that :)

kaushik-himself commented:
> * Should we update some distro examples with FiddleCube?
> * What's a good provider to test with?
The PR description contains details on how to use fiddlecube with the together distro. I can add the updated together template and distro to the PR.

> * Any doc update to reflect FiddleCube's availability to users?
Thanks for calling this out. Will update the relevant docs and add them to the PR.

> 
> Thanks!

kaushik-himself commented:
> Thanks for your contribution:
> 
> * Could you please update the CHANGELOG.md to reflect this new feature?
> * Should we update some distro examples with FiddleCube?
> * Any doc update to reflect FiddleCube's availability to users?
> * What's a good provider to test with?
> 
> Thanks!

- Updated distro examples with FiddleCube
- Updated docs
- Tested with the together provider

ashwinb reviewed (changes requested):
@kaushik-himself can you tell us why some of the existing distribution templates (e.g., nvidia, cerebras) have been changed to include this provider? In general, we do distro updates after some amount of consideration, testing, etc. I would ask that you remove changes to distros and merely use this PR for the provider contribution.

kaushik-himself commented:
> @kaushik-himself can you tell us why some of the existing distribution templates (e.g., nvidia, cerebras) have been changed to include this provider? In general, we do distro updates after some amount of consideration, testing, etc. I would ask that you remove changes to distros and merely use this PR for the provider contribution.

this was done as a response to @leseb's comment above, and my ignorance on what changes should be made to the templates/distros.

I will revert the changes to the distros and update the PR

ashwinb commented on llama_stack/providers/remote/safety/fiddlecube/fiddlecube.py:
```diff
@@ -0,0 +1,92 @@
+logger = logging.getLogger(__name__)
+
+
+class FiddlecubeSafetyAdapter(Safety, ShieldsProtocolPrivate):
```
Can you add a blurb about what the shield configuration should be? How should a user think about using guardrails from Fiddlecube? On that line, could you add a test also so it becomes somewhat self documenting? You can add a provider-focused test in `providers/tests/safety`

kaushik-himself replied:
added a blurb and added provider tests

ashwinb commented on llama_stack/providers/remote/safety/fiddlecube/fiddlecube.py:
```diff
@@ -0,0 +1,92 @@
+
+            logger.debug("Response:::", response.status_code)
+
+        # Check if the response is successful
```
nit: not a useful comment since code is self explanatory

ashwinb commented on llama_stack/providers/remote/safety/fiddlecube/fiddlecube.py:
```diff
@@ -0,0 +1,92 @@
+            logger.error(f"FiddleCube API error: {response.status_code} - {response.text}")
+            raise RuntimeError("Failed to run shield with FiddleCube API")
+
+        # Convert the response into the format RunShieldResponse expects
```
same

ashwinb commented on llama_stack/providers/remote/safety/fiddlecube/fiddlecube.py:
```diff
@@ -0,0 +1,92 @@
+
+        # Convert the response into the format RunShieldResponse expects
+        response_data = response.json()
+        logger.debug("Response data: %s", json.dumps(response_data, indent=2))
```
this is too much debug logging

ashwinb reviewed (changes requested):
A few comments

kaushik-himself commented on llama_stack/providers/tests/safety/conftest.py:
```diff
@@ -53,11 +53,19 @@
         id="remote",
         marks=pytest.mark.remote,
     ),
+    pytest.param(
```
@ashwinb I am working on adding a fiddlecube test here. However, I seem to have a test setup issue.

I tried running:
`pytest -s -m bedrock llama_stack/providers/tests/safety/test_safety.py` (also tried with `-m fiddlecube`)

Getting the error:
`ERROR llama_stack/providers/tests/safety/test_safety.py::TestSafety::test_run_shield[--bedrock] - ValueError: Model 'meta-llama/Llama-Guard-3-1B' is not available and no llama_model was specified in metadata. Please specify a llama_model in metadata or...`

ashwinb commented on llama_stack/providers/remote/safety/fiddlecube/fiddlecube.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class FiddlecubeSafetyAdapter(Safety, ShieldsProtocolPrivate):
+    """
```
👍

ashwinb reviewed (approved):
thx
